src/core/utils/math/bspline.hpp

```diff
@@ -0,0 +1,273 @@
+#ifndef UTILS_MATH_BSPLINE_HPP
+#define UTILS_MATH_BSPLINE_HPP
+
+#include "sqr.hpp"
+
+#include <cassert>
+#include <stdexcept>
+
+namespace Utils {
+template <int order, typename T = double> inline T bspline(int i, T x) {
```

Comment:
If you use template specializations instead of switches, the compiler can inline everything and most of the specializations can be marked `constexpr`.

Reply:
Codegen for this is fine. I'm currently working on refactoring this anyway, so I'll leave this unchanged here....